opm/autodiff/RedistributeDataHandles.hpp

```diff
@@ -0,0 +1,226 @@
+/*
+  Copyright 2015 Dr. Blatt - HPC-Simulation-Software & Services.
+  Coypright 2015 NTNU
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+
+#include <opm/autodiff/BlackoilPropsAdFromDeck.hpp>
+#include <opm/core/simulator/BlackoilState.hpp>
+
+namespace Opm
+{
+/// \brief a data handle to distribute the BlackoilState
+class BlackoilStateDataHandle
+{
+public:
+    /// \brief The data that we send.
+    typedef double DataType;
+    /// \brief Constructor.
+    /// \param sendGrid   The grid that the data is attached to when sending.
+    /// \param recvGrid   The grid that the data is attached to when receiving.
+    /// \param sendState  The state where we will retieve the values to be sent.
+    /// \parame recvState The state where we will store the received values.
+    BlackoilStateDataHandle(const Dune::CpGrid& sendGrid,
+                            const Dune::CpGrid& recvGrid,
+                            const BlackoilState& sendState,
+                            BlackoilState& recvState)
+        : sendGrid_(sendGrid), recvGrid_(recvGrid), sendState_(sendState), recvState_(recvState)
+    {}
+
+    bool fixedsize(int /*dim*/, int /*codim*/)
+    {
+        return false;
+    }
+
+    template<class T>
+    std::size_t size(const T& e)
+    {
+        if ( T::codimension == 0)
+        {
+            return 2 * sendState_.numPhases() +4+2*sendGrid_.numCellFaces(e.index());
+        }
+        else
+        {
+            OPM_THROW(std::logic_error, "Data handle can only be used for elements");
+        }
+    }
+
+    template<class B, class T>
```

Comment:
If the messages are not to very large, then yes. The buffer has maximum size. If the data that needs to be sent for the whole grid is larger than that then the messages will be split.